### ISSUE TITLE
Five second timeout for DNS lookup in 1.25.77

### ISSUE THREAD

**Issue by DentonGentry** [tailscale/tailscale#4686]
### What is the issue?

"ping example.com" freezes for 5 seconds before starting to ping in 1.25.77, which was not seen with 1.25.40. Notably, I believe 1.25.77 is the first unstable build to move MagicDNS to use netstack https://github.com/tailscale/tailscale/pull/4408

strace -tt ping apple.com shows:
```
03:07:31.543860 setsockopt(5, SOL_IP, IP_RECVERR, [1], 4) = 0
03:07:31.543953 connect(5, {sa_family=AF_INET, sin_port=htons(53), sin_addr=inet_addr("100.100.100.100")}, 16) = 0
03:07:31.544076 poll([{fd=5, events=POLLOUT}], 1, 0) = 1 ([{fd=5, revents=POLLOUT}])
03:07:31.544182 sendmmsg(5, [{msg_hdr={msg_name=NULL, msg_namelen=0, msg_iov=[{iov_base="\354\325\1\0\0\1\0\0\0\0\0\0\5apple\3com\0\0\1\0\1", iov_len=27}], msg_iovlen=1, msg_controllen=0, msg_flags=0}, msg_len=27}, {msg_hdr={msg_name=NULL, msg_namelen=0, msg_iov=[{iov_base="Y\332\1\0\0\1\0\0\0\0\0\0\5apple\3com\0\0\34\0\1", iov_len=27}], msg_iovlen=1, msg_controllen=0, msg_flags=0}, msg_len=27}], 2, MSG_NOSIGNAL) = 2
03:07:31.544363 poll([{fd=5, events=POLLIN}], 1, 5000) = 1 ([{fd=5, revents=POLLIN}])
03:07:31.577430 ioctl(5, FIONREAD, [43]) = 0
03:07:31.577562 recvfrom(5, "\354\325\201\200\0\1\0\1\0\0\0\0\5apple\3com\0\0\1\0\1\300\f\0\1\0"..., 2048, 0, {sa_family=AF_INET, sin_port=htons(53), sin_addr=inet_addr("100.100.100.100")}, [28->16]) = 43
03:07:31.577748 poll([{fd=5, events=POLLIN}], 1, 4966

...this is where the ping paused for a few seconds. I hit return a few times...

) = 0 (Timeout)
03:07:36.545435 poll([{fd=5, events=POLLOUT}], 1, 0) = 1 ([{fd=5, revents=POLLOUT}])
03:07:36.545623 sendto(5, "\354\325\1\0\0\1\0\0\0\0\0\0\5apple\3com\0\0\1\0\1", 27, MSG_NOSIGNAL, NULL, 0) = 27
03:07:36.545884 poll([{fd=5, events=POLLIN}], 1, 5000) = 1 ([{fd=5, revents=POLLIN}])
03:07:36.551934 ioctl(5, FIONREAD, [43]) = 0
03:07:36.552069 recvfrom(5, "\354\325\201\200\0\1\0\1\0\0\0\0\5apple\3com\0\0\1\0\1\300\f\0\1\0"..., 2048, 0, {sa_family=AF_INET, sin_port=htons(53), sin_addr=inet_addr("100.100.100.100")}, [28->16]) = 43
03:07:36.552237 poll([{fd=5, events=POLLOUT}], 1, 4993) = 1 ([{fd=5, revents=POLLOUT}])
03:07:36.552354 sendto(5, "Y\332\1\0\0\1\0\0\0\0\0\0\5apple\3com\0\0\34\0\1", 27, MSG_NOSIGNAL, NULL, 0) = 27
03:07:36.552575 poll([{fd=5, events=POLLIN}], 1, 4992) = 1 ([{fd=5, revents=POLLIN}])
03:07:36.558714 ioctl(5, FIONREAD, [96]) = 0
03:07:36.558899 recvfrom(5, "Y\332\201\200\0\1\0\0\0\1\0\0\5apple\3com\0\0\34\0\1\300\f\0\6\0"..., 65536, 0, {sa_family=AF_INET, sin_port=htons(53), sin_addr=inet_addr("100.100.100.100")}, [28->16]) = 96
03:07:36.559126 close(5)                = 0
```

### OS

Linux

### OS version

Debian 11.0

### Tailscale version

1.25.77

**Comment by twitchyliquid64** [comment]
I'm able to reproduce with the ping from inetutils. Whats strange to me is that its clear that theres a reply coming back to the process ... hmmm ...

I presume theres some wierd semantics with `AI_CANONNAME`:

https://github.com/guillemj/inetutils/blob/cf091304a3218efa3bc267b483cad71b6f65c24e/ping/libping.c#L307-L317

I'll work on a small repro and figure out whats going on with that hint.

**Comment by twitchyliquid64** [comment]
Looks like Chrome ran into this as well, it doesn't seem to be related to hints but rather to the network configuration on your host. Can you share your `ifconfig` / if you have ipv6 working @DentonGentry ?

I'm able to repro running the [test tool](https://bugs.chromium.org/p/chromium/issues/attachment?aid=26310&signed_aid=Udsg6D022CHbffC7Egwleg==) from the chrome bug, but only the first time after bringing up `tailscaled`:

```shell
[xxx@tslptp:/tmp]$ g++ -o test ~/Downloads/getaddrinfo_experiment.cc && ./test apple.com
-------------------------------------------------------
getaddrinfo(host="apple.com", hints=NULL):
  Duration: 10.0407 seconds
  Succeeded with:
    0: 17.253.144.10
    1: 17.253.144.10
    2: 17.253.144.10
-------------------------------------------------------
getaddrinfo(host="apple.com", hints={ai_family=AF_UNSPEC, ai_flags=0}):
  Duration: 0.0256169 seconds
  Succeeded with:
    0: 17.253.144.10
-------------------------------------------------------
getaddrinfo(host="apple.com", hints={ai_family=AF_UNSPEC, ai_flags=AI_CANONNAME}):
  Duration: 0.025115 seconds
  Succeeded with:
    0: 17.253.144.10
-------------------------------------------------------
getaddrinfo(host="apple.com", hints={ai_family=AF_UNSPEC, ai_flags=AI_ADDRCONFIG}):
  Duration: 0.0259302 seconds
  Succeeded with:
    0: 17.253.144.10
-------------------------------------------------------
getaddrinfo(host="apple.com", hints={ai_family=AF_UNSPEC, ai_flags=AI_ADDRCONFIG | AI_CANONNAME}):
  Duration: 0.0254471 seconds
  Succeeded with:
    0: 17.253.144.10
-------------------------------------------------------
getaddrinfo(host="apple.com", hints={ai_family=AF_UNSPEC, ai_flags=AI_ADDRCONFIG | AI_V4MAPPED}):
  Duration: 0.0271342 seconds
  Succeeded with:
    0: 17.253.144.10
-------------------------------------------------------
getaddrinfo(host="apple.com", hints={ai_family=AF_INET, ai_flags=0}):
  Duration: 0.0150819 seconds
  Succeeded with:
    0: 17.253.144.10
-------------------------------------------------------
getaddrinfo(host="apple.com", hints={ai_family=AF_INET, ai_flags=AI_CANONNAME}):
  Duration: 0.0120411 seconds
  Succeeded with:
    0: 17.253.144.10
-------------------------------------------------------
getaddrinfo(host="apple.com", hints={ai_family=AF_INET, ai_flags=AI_ADDRCONFIG}):
  Duration: 0.0125809 seconds
  Succeeded with:
    0: 17.253.144.10
-------------------------------------------------------
getaddrinfo(host="apple.com", hints={ai_family=AF_INET, ai_flags=AI_ADDRCONFIG | AI_CANONNAME}):
  Duration: 0.0127499 seconds
  Succeeded with:
    0: 17.253.144.10
-------------------------------------------------------
getaddrinfo(host="apple.com", hints={ai_family=AF_INET, ai_flags=AI_ADDRCONFIG | AI_V4MAPPED}):
  Duration: 0.013128 seconds
  Succeeded with:
    0: 17.253.144.10
-------------------------------------------------------
getaddrinfo(host="apple.com", hints={ai_family=AF_INET6, ai_flags=0}):
  Duration: 0.340493 seconds
  ERROR: -5: No address associated with hostname
-------------------------------------------------------
getaddrinfo(host="apple.com", hints={ai_family=AF_INET6, ai_flags=AI_CANONNAME}):
  Duration: 0.122767 seconds
  ERROR: -5: No address associated with hostname
-------------------------------------------------------
getaddrinfo(host="apple.com", hints={ai_family=AF_INET6, ai_flags=AI_ADDRCONFIG}):
  Duration: 0.133276 seconds
  ERROR: -5: No address associated with hostname
-------------------------------------------------------
getaddrinfo(host="apple.com", hints={ai_family=AF_INET6, ai_flags=AI_ADDRCONFIG | AI_CANONNAME}):
  Duration: 0.123943 seconds
  ERROR: -5: No address associated with hostname
-------------------------------------------------------
getaddrinfo(host="apple.com", hints={ai_family=AF_INET6, ai_flags=AI_ADDRCONFIG | AI_V4MAPPED}):
  Duration: 0.134037 seconds
  Succeeded with:
    0: ::ffff:17.253.144.10

[xxx@tslptp:/tmp]$ g++ -o test ~/Downloads/getaddrinfo_experiment.cc && ./test apple.com
-------------------------------------------------------
getaddrinfo(host="apple.com", hints=NULL):
  Duration: 0.000183821 seconds
  Succeeded with:
    0: 17.253.144.10
    1: 17.253.144.10
    2: 17.253.144.10
-------------------------------------------------------
getaddrinfo(host="apple.com", hints={ai_family=AF_UNSPEC, ai_flags=0}):
  Duration: 0 seconds
  Succeeded with:
    0: 17.253.144.10
-------------------------------------------------------
getaddrinfo(host="apple.com", hints={ai_family=AF_UNSPEC, ai_flags=AI_CANONNAME}):
  Duration: 9.53674e-07 seconds
  Succeeded with:
    0: 17.253.144.10
-------------------------------------------------------
getaddrinfo(host="apple.com", hints={ai_family=AF_UNSPEC, ai_flags=AI_ADDRCONFIG}):
  Duration: 0 seconds
  Succeeded with:
    0: 17.253.144.10
-------------------------------------------------------
getaddrinfo(host="apple.com", hints={ai_family=AF_UNSPEC, ai_flags=AI_ADDRCONFIG | AI_CANONNAME}):
  Duration: 0 seconds
  Succeeded with:
    0: 17.253.144.10
-------------------------------------------------------
getaddrinfo(host="apple.com", hints={ai_family=AF_UNSPEC, ai_flags=AI_ADDRCONFIG | AI_V4MAPPED}):
  Duration: 9.53674e-07 seconds
  Succeeded with:
    0: 17.253.144.10
-------------------------------------------------------
getaddrinfo(host="apple.com", hints={ai_family=AF_INET, ai_flags=0}):
  Duration: 0.0135078 seconds
  Succeeded with:
    0: 17.253.144.10
-------------------------------------------------------
getaddrinfo(host="apple.com", hints={ai_family=AF_INET, ai_flags=AI_CANONNAME}):
  Duration: 1.90735e-06 seconds
  Succeeded with:
    0: 17.253.144.10
-------------------------------------------------------
getaddrinfo(host="apple.com", hints={ai_family=AF_INET, ai_flags=AI_ADDRCONFIG}):
  Duration: 8.10623e-06 seconds
  Succeeded with:
    0: 17.253.144.10
-------------------------------------------------------
getaddrinfo(host="apple.com", hints={ai_family=AF_INET, ai_flags=AI_ADDRCONFIG | AI_CANONNAME}):
  Duration: 0 seconds
  Succeeded with:
    0: 17.253.144.10
-------------------------------------------------------
getaddrinfo(host="apple.com", hints={ai_family=AF_INET, ai_flags=AI_ADDRCONFIG | AI_V4MAPPED}):
  Duration: 9.53674e-07 seconds
  Succeeded with:
    0: 17.253.144.10
-------------------------------------------------------
getaddrinfo(host="apple.com", hints={ai_family=AF_INET6, ai_flags=0}):
  Duration: 0 seconds
  ERROR: -2: Name or service not known
-------------------------------------------------------
getaddrinfo(host="apple.com", hints={ai_family=AF_INET6, ai_flags=AI_CANONNAME}):
  Duration: 0 seconds
  ERROR: -2: Name or service not known
-------------------------------------------------------
getaddrinfo(host="apple.com", hints={ai_family=AF_INET6, ai_flags=AI_ADDRCONFIG}):
  Duration: 0 seconds
  ERROR: -2: Name or service not known
-------------------------------------------------------
getaddrinfo(host="apple.com", hints={ai_family=AF_INET6, ai_flags=AI_ADDRCONFIG | AI_CANONNAME}):
  Duration: 0 seconds
  ERROR: -2: Name or service not known
-------------------------------------------------------
getaddrinfo(host="apple.com", hints={ai_family=AF_INET6, ai_flags=AI_ADDRCONFIG | AI_V4MAPPED}):
  Duration: 0 seconds
  Succeeded with:
    0: ::ffff:17.253.144.10
```

**Comment by DentonGentry** [comment]
`ip addr show` is pasted below. It looks like it does not have working IPv6 (other systems in the house do, but not this one I guess).

```
# ip addr show
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever
2: eth0@if1135: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default qlen 1000
    link/ether 0e:6b:cd:ae:a2:9a brd ff:ff:ff:ff:ff:ff link-netnsid 0
    inet 10.1.10.12/24 brd 10.1.10.255 scope global dynamic eth0
       valid_lft 3974sec preferred_lft 3974sec
    inet6 fe80::c6b:cdff:feae:a29a/64 scope link 
       valid_lft forever preferred_lft forever
5: tailscale0: <POINTOPOINT,MULTICAST,NOARP,UP,LOWER_UP> mtu 1280 qdisc pfifo_fast state UNKNOWN group default qlen 500
    link/none 
    inet 100.104.59.4/32 scope global tailscale0
       valid_lft forever preferred_lft forever
    inet6 fd7a:115c:a1e0:ab12:4843:cd96:6268:3b04/128 scope global 
       valid_lft forever preferred_lft forever
    inet6 fe80::1f61:3f56:7847:3555/64 scope link stable-privacy 
       valid_lft forever preferred_lft forever
```

**Comment by twitchyliquid64** [comment]
Okay, I think I understand this. glibc delegates out to its own library for resolution, `libresolv.so`:

```
openat(AT_FDCWD, "/nix/store/qjgj2642srlbr59wwdihnn66sw97ming-glibc-2.33-123/lib/libresolv.so.2",
        O_RDONLY|O_CLOEXEC) = 3
```

Which, if nscd is not running on your system, looks at your network configuration to determine if you have any cheeky ipv6 and hence if you need a query for quad-AAAA records:

```
socket(AF_NETLINK, SOCK_RAW|SOCK_CLOEXEC, NETLINK_ROUTE) = 3
bind(3, {sa_family=AF_NETLINK, nl_pid=0, nl_groups=00000000}, 12) = 0
getsockname(3, {sa_family=AF_NETLINK, nl_pid=940997, nl_groups=00000000}, [12]) = 0
```

After parsing `resolv.conf`, it connects to the configured resolver over UDP and yeets **two** queries down the same socket, one for a A record and one for a AAAA record:

```
connect(3, {sa_family=AF_INET, sin_port=htons(53), sin_addr=inet_addr("100.100.100.100")}, 16) = 0
poll([{fd=3, events=POLLOUT}], 1, 0)    = 1 ([{fd=3, revents=POLLOUT}])
sendmmsg(3, [
    {msg_hdr={msg_name=NULL, msg_namelen=0, msg_iov=[
      {iov_base="\270\"\1\0\0\1\0\0\0\0\0\1\5apple\3com\0\0\1\0\1\0\0)\4\260"..., iov_len=38}], msg_iovlen=1, msg_controllen=0, msg_flags=0}, msg_len=38},
    {msg_hdr={msg_name=NULL, msg_namelen=0, msg_iov=[
      {iov_base="p \1\0\0\1\0\0\0\0\0\1\5apple\3com\0\0\34\0\1\0\0)\4\260"..., iov_len=38}], msg_iovlen=1, msg_controllen=0, msg_flags=0}, msg_len=38}
      ], 2, MSG_NOSIGNAL) = 2
```

This doesn't vibe with the netstack implementation, which assumes that one UDP socket = one DNS query (the pre-netstack implementation did not understand the concept of different UDP sockets, so we never hit this).

Regardless of whether glibc is doing the right thing or not (its not well-defined in the spec), the fix is to permit multiplexing DNS queries down a single UDP socket.

**Comment by DentonGentry** [comment]
There is another mention of 5s DNS delay in https://github.com/tailscale/corp/issues/3060, but on macOS and Tailscale 1.18 I think.

**Comment by aleon-jsjiang** [comment]
> There is another mention of 5s DNS delay in https://github.com/tailscale/corp/issues/3060, but on macOS and Tailscale 1.18 I think.

Hi @DentonGentry,

I suppose that's an internally tracked issue, but I am encountering exactly 5s MagicDNS delays on macOS, Tailscale 1.28 (App Store version). I realize this issue is closed but I found your comment and wanted to share as it sounds like the same problem - I can open a new issue if that would be better.

```
time dscacheutil -q host -a name {host name}                                                                        
name: {host name}.{domain}.beta.tailscale.net
ip_address: 100.93.12.46

dscacheutil -q host -a name {host name}  0.00s user 0.00s system 0% cpu 5.014 total
```

If I lookup by the fqdn rather than machine name, resolution is quick:
```
time dscacheutil -q host -a name {host name}.{domain}.beta.tailscale.net                                     
name: {host name}.{domain}.beta.tailscale.net
ip_address: 100.93.12.46

dscacheutil -q host -a name   0.00s user 0.00s system 36% cpu 0.016 total
```

This delay is noticeable via browser, `curl`, `nc`, etc. when using only the machine name. Interestingly, using `ping` resolves quickly - I'm not sure if there is some caching going behind the scenes there.

**Comment by DentonGentry** [comment]
Please open a new issue.
